lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+    """
+    tty.debug('Bootstrapping {0} compiler for {1}'
+              .format(pkg.spec.compiler, pkg.unique_id))
+    compilers = spack.compilers.compilers_for_spec(
+        pkg.spec.compiler, arch_spec=pkg.spec.architecture)
+    if not compilers:
+        dep = spack.compilers.pkg_spec_for_compiler(pkg.spec.compiler)
+        dep.architecture = pkg.spec.architecture
+        # concrete CompilerSpec has less info than concrete Spec
+        # concretize as Spec to add that information
+        dep.concretize()
+        packages = [(s.package, False) for
+                    s in dep.traverse(order='post', root=False)]
+        packages.append((dep.package, True))
+        return packages
+
+    return []
+
+
+def _hms(seconds):
+    """
+    Convert seconds to hours, minutes, seconds
+
+    Args:
+        seconds (int): time to be converted in seconds
+
+    Return:
+        (str) String representation of the time as #h #m #.##fs
+    """
+    m, s = divmod(seconds, 60)
+    h, m = divmod(m, 60)
+
+    parts = []
+    if h:
+        parts.append("%dh" % h)
+    if m:
+        parts.append("%dm" % m)
+    if s:
+        parts.append("%.2fs" % s)
+    return ' '.join(parts)
+
+
+def _installed_from_cache(pkg, cache_only, explicit):
+    """
+    Install the package from binary cache
+
+    Args:
+        pkg (spack.package.PackageBase): the package to install from the binary
+            cache
+        cache_only (bool): only install from binary cache
+        explicit (bool): ``True`` if installing the package was explicitly
+            requested by the user, otherwise, ``False``
+
+    Return:
+        (bool) ``True`` if the package was installed from binary cache,
+            ``False`` otherwise
+    """
+    if _try_install_from_binary_cache(pkg, explicit):
+        tty.debug('Successfully installed {0} from binary cache'
+                  .format(pkg.unique_id))
+        _print_installed_pkg(pkg.spec.prefix)
+        spack.hooks.post_install(pkg.spec)
+        return True
+    elif cache_only:
+        tty.die('No binary for {0} found when cache-only specified'
+                .format(pkg.unique_id))
+
+    tty.debug('No binary for {0} found: installing from source'
+              .format(pkg.unique_id))
+    return False
+
+
+def _print_installed_pkg(message):
+    """
+    Output a message with a package icon.
+
+    Args:
+        message (str): message to be output
+    """
+    cwrite('@*g{[+]} ')
+    print(message)
+
+
+def _process_external_package(pkg, explicit):
+    """
+    Helper function to run post install hooks and register external packages.
+
+    Args:
+        pkg (Package): the external package
+        explicit (bool): if the package was requested explicitly by the user,
+            ``False`` if it was pulled in as a dependency of an explicit
+            package.
+    """
+    # TODO: Should this error out?
+    if not pkg.spec.external:
+        return
```

Comment:
I think this should error in some way, since if we call this function a prerequisite is that the package is external. Maybe:
```python
assert pkg.spec.external, error_message
```
would do as an error?

Reply:
Sounds reasonable.  Thanks.